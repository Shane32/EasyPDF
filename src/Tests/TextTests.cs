using System.Drawing;
using System.Text;
using Font = Shane32.EasyPDF.Font;

namespace Tests;

public class TextTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public TextTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void PositioningTests()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
