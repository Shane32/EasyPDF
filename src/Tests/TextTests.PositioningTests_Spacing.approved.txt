%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 895/Filter/FlateDecode>>stream
x\x9C\x8D\x96Ko\xD4@\x0C\x80\xEF\xF9\x159\xC2a\x83=\x99<\x86\x1BE\x80\x84\xB8P\xF6\x868\xA0v[Zu[Z*\xF1\xF7\x99\xD8\xE3\xC7\x94m7\xEAc\xDD\x89\xEDq>%_\xE7\xBE9\xD96\xFD\xD8N\xC3\xD8n\xCF\x9B\x0F\xDB\xE6k\x83-\xE4\xAF\xB6S\xC8\xDF\xD0\x9E\xED\x1Bh?\xE7\x9F\xEB\x06:l\xFF6\xDF\x7F\xE4\xEB\xE7P\xDE\xE9\xA7\x12<\\\x1EN\x0B\x01\xBA!G\xFB\x12\x85.\xA66\xE0\xBC|\xC4\xE5o\x1C\xE9\xF3\xAC\x09\xD8w\x03\x96E\xB4T\xA4\xFA\xE5\xFA\x12m\xCA*\xE5n\xA4\xC3FZP_^\xA6\xEDJz\x19\xE2\xAC\xF9\xD6H\xC3\xBD\x8EvC\xAB\xA5\xC9\xBE\x91\x89\x96\xE5\xFB\x05\xD0\x9B\x8F\xD8"\xB4\xDB\x0Bc\xB3\xE4L\x1D\xCE\xEDv\xEF\xEF\x1F\x16\x8A\xAF\xDE\xDD\\]\xDE\xEEw\xB7\x8Fo\xDB/\xBB\x8B\xC7\xED\xDD\xEF\xD7\xDBk\x86\xCB[\xE6:\xC5\x01\x8EG\x08\xE5v(P"e\x19]:r\x17\x81\x82\x03\xE5q\xFA\xF2\x9B\xBAP\xA0Px\x99\xF7\xE7t\x99\xC5\xA8\xB8\xC1r(\\\xA8\x92\xB9\xD0(\xC7\xC0\x84\xDC*\xAD!\xF3>G\xBB\x87'p\xF2\xE3(3\xF4\xB3\x7FX@\x9E\x16\xA8\x1E\x17P8\x92\x8E\xDCE\xE0\xF4\xBD\x83\xD3c\xE9B\x81\xC2\xE1e\xDE\xB478\xD4E\xE0\xF8\xC1F\x85C\x95\xE5\xA1\x81\x15pr\xE9
2'?\xFF\xECn\xAEnwO\xD8\x0CQG\x18F\xC7f\x98\xCB]Q\xA0l\xCA2\xBAt\xE4.\xC2f@\xC7&&!\x9C<\x1B^\xE6M\xD1\xD8P\x17a\xE3\x07\x8B\xCA\x86*\x99\x8Dr\x8C\xCD\x00\x1D\xACz\xA5N\xEE\x1E\x1F\xEF\xF6O\xE0d_\xC9\x0CStp\xA6\xB1\xDC\x16\x05
\xA7,\xA3KG\xEE"p\xC6\xE4\xE0\x8CS\xE9B\x81\xC2\xE1e\xDA\xB4\xA4\xCB,\x06\xC7\x0F\x16\x14\x0EU2\x1C\x1A\xE5%8\x98\x93\xFB\xA9\x9D\xD2
\xE5\xF0K\xF5\xBFt\x92)/y\x09'qh\xAA4\x9C\xCC\xC3\xC9\x898\x99\x89\xE7\xC9\xE1\x99\x87\xD2\x85\x02\xC5\xC3\xCB\xB4iI\x97Y\x0C\x8F\x1F\xCCdL\x95\x8C'\x1D\xB31\x8E\xD0\xA5\xB9M\xFD
\xEF0\x9E\x83\xE6Ap\xFAC/fD13GJI.\xA0/\xC1\xD2K\xED\x0C\x95\x9EA\xFD\x0C\xB5\xA0\xC1\x19\x1A\xBC\xA2\xA1r\xB4\x9F\x12\x9C\xA5\xC14\xCDc\xBD\x88,\x0BaFj\xB0
\xD83B\xC2`R\xC4\xE0u\x8D\xBD\xF8\x9A#C\xD6\x9B\xB1\xB5\x04K/E\x16\xBC\xB41\x88\xB592d\xC1\xBC-%:\x97CVMi\xEE\xE6\xF2\x82\xAC?f\xEF\xE5)\xC3\x98\x8B\xC2
K\x15j\x87<\x85\xD1d\x89\xD1k\x1C\xA3x\x9C#c\x16\xCD\xE4Z\x82\xA5\x972\x8BX\x9D\x02\x92\xF2\xAFt^.\x94\xED\x9D\xD0\xB9\x972\xAB\xA6\x8C\xEE0`R\xE7\xB1^d\x16\x86..O\x1B\xAE0\xD7\xE9\xD5\xE5\xAFC\xA7\xA5\xD1\xFC\x89\xA37;\x8E\xA2v\x8E\x8C\xD7hr\xD7\x12,\xBD\xEC\xD0\xE4\xFD\x8E\x83\x08\x9E#\xE35\x98\xE2\xA5D\xE7r\xBC\xAA)M\xF3\\^x\x8DGE\x9F_\xDC\xB1\xCF\x89\xC3
\x95\x11\xAF\xC3&\x9B\xC1\x1D\xE4\xBC\xEDq\x16\xDDsd\xC8f\x13\xBE\x96`\xE9\xA5\xC8&\xEF|\x9CD\xFA\x1C\x19\xB2\xC9\xB4/%:\x97?m\xFA)M\xFD\\^\x90\xCDG\xE5\x0F\xB1\x0B\x89\x1A\xAC\xE1\xF5\x9C\xC8\x92?\x93W\x87r\xD0S9\xD4\xC7r\xF0\xE7r\xEF\xFE\xE4\xDC\x9F*\xF7'u\x7F\xAA\xDD\x9F\x9C\xFB\x93w\x7F\xAA\xDC\xEF\xA6\\b%\x96\xDC\x11\x1D\x8E\xBA?\xFF\xA3JKQ\\!2\x86V{\xEC\x1F\x97/\xE4\xEB
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001134 00000 n 
0000000015 00000 n 
0000001222 00000 n 
0000000977 00000 n 
0000001287 00000 n 
0000001332 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 